apps/aechannel/src/aesc_fsm.erl

```diff
@@ -2894,42 +3150,43 @@ process_update_error(Reason, From, D) ->
     ?LOG_CAUGHT(Reason),
     keep_state(D, [{reply, From, {error, Reason}}]).
 
-check_closing_event(#data{on_chain_id = ChanId} = D) ->
-    case aec_chain:get_channel(ChanId) of
-        {ok, Channel} ->
-            case aesc_channels:is_active(Channel) of
-                true ->
-                    check_closing_event_(Channel, D);
-                false ->
-                    {error, not_active}
-            end;
-        {error, _} = Error ->
-            Error
-    end.
-
-check_closing_event_(Ch, #data{state = St}) ->
-    Height = cur_height(),
+check_closing_event(Info, D) ->
+    aec_db:dirty(fun() ->
```

Comment:
I see the only DB operation inside this dirty context is `aec_chain:get_header` that anyway adds transactional context. If optimized DB read here is really necessary, I wonder: why not removing need for handling transaction inside dirty context by cutting one layer of abstraction and perform read in dirty activity directly? Something like:

```erlang
check_closing_event(Info, D) ->
    {value, Hdr} = aec_db:find_header(async_dirty, maps:get(block_hash, Info)), %% TODO `aec_db:find_header/2`.
    try check_closing_event_(maps:get(tx, Info), maps:get(channel, Info), Hdr, D)
    catch
        error:E ->
            ?LOG_CAUGHT(E),
            error(E)
    end.
```
If need be, read can be made conditional to `aesc_channels:is_solo_closing(Ch)`.

Reply:
I see that calls to `aec_db:ensure_activity(async_dirty, ...)` are out of scope for this ^^^ comment, as those follow other code path not attempting to nest transaction inside dirty activity.

Reply:
> I see the only DB operation inside this dirty context is `aec_chain:get_header` that anyway adds transactional context. 

Actually, this didn't use to be true, until PR #2477 changed the `?t(Expr)` macro in `aec_db.erl`. I believe this was a mistake. The `aec_db` read operations shouldn't add a transaction context if there is already a transaction _or_ non-transaction context.

Reply:
As this PR is not based on latest master, when merged this PR will keep the `?t` macro definition in master.

Reply:
The `?t` macro is now aligned to master in 6ca8c386d